apis/python/src/tiledbsoma/experiment_query/query.py

```diff
@@ -0,0 +1,483 @@
+from ..dataframe import DataFrame as SOMADataFrame
+from ..sparse_nd_array import SparseNDArrayRead
+
+if TYPE_CHECKING:
+    from ..experiment import Experiment
+from ..sparse_nd_array import SparseNDArray as SOMASparseNDArray
+from .anndata import _make_anndata
+from .axis import AxisQuery
+from .eq_types import AxisColumnNames, ExperimentAxisQueryReadArrowResult
+
+AxisJoinIds = TypedDict(
+    "AxisJoinIds",
+    {
+        "obs": Optional[pa.Array],
+        "var": Optional[pa.Array],
+    },
+)
+
+MatrixAxisQuery = TypedDict(
+    "MatrixAxisQuery",
+    {
+        "obs": AxisQuery,
+        "var": AxisQuery,
+    },
+)
+
+
+class ExperimentAxisQuery(ContextManager["ExperimentAxisQuery"]):
+    """
+    ExperimentAxisQuery allows easy selection and extraction of data from a single soma.Measurement
+    in a soma.Experiment, by obs/var (axis) coordinates and/or value filter [lifecycle: experimental].
+
+    The primary use for this class is slicing Experiment ``X`` layers by obs or var value and/or coordinates.
+    Slicing on SparseNDArray ``X`` matrices is supported; DenseNDArray is not supported at this time.
+
+    IMPORTANT: this class is not thread-safe.
+
+    IMPORTANT: this query class assumes it can store the full result of both axis dataframe
+    queries in memory, and only provides incremental access to the underlying X NDArray. API
+    features such as `n_obs` and `n_vars` codify this in the API.
+
+    IMPORTANT: you must call `close()` on any instance of this class in order to release
+    underlying resources. The ExperimentAxisQuery is a context manager, and it is recommended
+    that you use the following pattern to make this easy and safe:
+    ```
+        with ExperimentAxisQuery(...) as query:
+            ...
+    ```
+    """
+
+    def __init__(
+        self,
+        experiment: "Experiment",
+        measurement_name: str,
+        *,
+        obs_query: Optional[AxisQuery] = None,
+        var_query: Optional[AxisQuery] = None,
+    ):
+        if not experiment.exists():
+            raise ValueError("Experiment does not exist")
+        if measurement_name not in experiment.ms:
+            raise ValueError("Measurement does not exist in the experiment")
+
+        self.experiment = experiment
+        self.measurement_name = measurement_name
+
+        self._matrix_axis_query: MatrixAxisQuery = {
+            "obs": obs_query if obs_query is not None else AxisQuery(),
+            "var": var_query if var_query is not None else AxisQuery(),
+        }
+        self._joinids: AxisJoinIds = {
+            "obs": None,
+            "var": None,
+        }
+        self._indexer: "AxisIndexer" = AxisIndexer(self)
+
+        self._threadpool_: Optional[concurrent.futures.ThreadPoolExecutor] = None
+
+    def close(self) -> None:
+        """
+        Clean up and close all resources. This must be called or the thread pool
+        will not be released.
+        """
+        if self._threadpool_ is not None:
```

Comment:
nit: this can be `if self._threadpool_` since all threadpools are truthy

Reply:
Actually, the suggested method fails mypy static analysis, whereas the explicit None test passes.  Leaving as is.